From 256c4293629788772773cfd5334e47b9c030865d Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Date: Thu, 4 Apr 2024 11:12:59 +0530
Subject: [PATCH] wifi: ath12k: parse and enable beacon protection for non-tx
 BSS

Currently, beacon protection support is checked for tx and non tx
bss based on transmitted profile bss, which limits the use of beacon
protection efficiently in both tx and non tx BSS.

Fix this limitation by enabling the beacon protection when
Non-transmitted BSS's Beacon Protection Support bit is set.

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1789,6 +1789,7 @@ static bool ath12k_mac_setup_bcn_tmpl_no
 	struct ath12k_vif *ahvif = arvif->ahvif;
 	struct ieee80211_mgmt *mgmt;
 	const u8 *ies, *profile, *next_profile;
+	const struct element *elem;
 	int ies_len;
 
 	if (arvif == tx_arvif)
@@ -1826,6 +1827,11 @@ static bool ath12k_mac_setup_bcn_tmpl_no
 				profile_len -= 5;
 				profile = profile + 5;
 
+				elem = cfg80211_find_elem(WLAN_EID_EXT_CAPABILITY, profile, profile_len);
+				if (elem && elem->datalen >= 11 &&
+				    (elem->data[10] & WLAN_EXT_CAPA11_BCN_PROTECT))
+					arvif->beacon_prot = tx_arvif->beacon_prot = true;
+
 				if (cfg80211_find_ie(WLAN_EID_RSN, profile,
 						     profile_len))
 					ahvif->rsnie_present = true;
